src/main/java/hudson/plugins/deploy/DeployPublisher.java

```diff
@@ -47,17 +46,43 @@ public DeployPublisher(List<ContainerAdapter> adapters, String war, String conte
         this.contextPath = contextPath;
     }
 
+    @SuppressFBWarnings("NP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE")
     @Override
+    @Deprecated
     public boolean perform(AbstractBuild<?,?> build, Launcher launcher, BuildListener listener) throws InterruptedException, IOException {
-        if (build.getResult().equals(Result.SUCCESS) || onFailure) {
-            for (FilePath warFile : build.getWorkspace().list(this.war)) {
-                for (ContainerAdapter adapter : adapters)
-                    if (!adapter.redeploy(warFile, contextPath, build, launcher, listener))
-                        build.setResult(Result.FAILURE);
+        perform(false, build, build.getWorkspace(), launcher, listener);
+        return true;
+    }
+
+    @Override
+    public void perform(@Nonnull Run<?, ?> run, @Nonnull FilePath workspace, @Nonnull Launcher launcher, @Nonnull TaskListener listener) throws InterruptedException, IOException {
+        perform(true, run, workspace, launcher, listener);
+    }
+
+    private void perform (boolean fromWorkFlow, @Nonnull Run<?, ?> run, @Nonnull FilePath workspace, @Nonnull Launcher launcher, @Nonnull TaskListener listener) throws InterruptedException, IOException {
+        if (fromWorkFlow || onFailure || Result.SUCCESS.equals(run.getResult())) {
+            if (!workspace.exists()) {
+                listener.getLogger().println("[DeployPublisher][ERROR] Workspace not found");
+                throw new FileNotFoundException("Workspace not found");
             }
-        }
 
-        return true;
+            FilePath[] wars = workspace.list(this.war);
+            if (wars.length == 0) {
+                listener.getLogger().printf("[DeployPublisher][WARN] No wars found. Deploy aborted. %n");
+                return;
+            }
+            listener.getLogger().printf("[DeployPublisher][INFO] Attempting to deploy %d war file(s)%n", wars.length);
+
+            for (FilePath warFile : wars) {
+                for (ContainerAdapter adapter : adapters) {
+                    if (!adapter.redeploy(warFile, contextPath, run, launcher, listener)) {
```

Comment:
If one of the deployers fails, all others won't be tried. I see this comes from the original code, but I would check if there are open bugs for it.

Another potential risk is a RunTime exception being thrown by a bogus extension implementation. BKM is to not trust extensions and to catch these exceptions and rethrow as a checked exception in order to avoid impact on the plugin logic